### PR TITLE
arc: Add support for cooperative sharing of VPX vector registers

### PR DESCRIPTION
**Pull request by peter-mitsis**
Saving the VPX vector registers on each context switch would incur a significant penalty that would likely negate much of their benefit. This set of commits implements a public API that models them as a critical resource and allows multiple threads access to safely access them provided that they follow the rules of acquiring the lock (via arc_vpx_lock) before using them, and releasing the lock (via arc_vpx_unlock) when done using them.

### CODE REVIEW

**Comment by KeHIntel** [comment]
Add @qipengzha and @yeweizy to help review this PR

**Comment by ruuddw** [comment]
Thanks, I really like the idea to use a semaphore to protect cooperative sharing of the DSP vector context. For scalar context, there already is support for indicating per task if the context should be saved or not, but for VPX vector that context is too large to use that mechanism. Few questions/discussion topics:

- why couple it to the ARC_DSP feature? ARC_DSP is for HS type scalar DSP, not VPX vectors. Isn't this an independent feature?
- any reason for the design choice to always arc_vpx_lock() the current CPU, but allowing the arc_vpx_unlock() to unlock any id? It may be slightly more efficient, but could also be misused (other cpu unlocking it). How about a more symmetrical interface, with either both lock/unlock getting an id parameter, or non?
- the assert for cpu-mask checking (good) also has ISR checking; is it allowed/supported to call these functions from ISR context?

File: tests/arch/arc/arc_vpx_lock/src/main.c [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+		k_thread_suspend(data->thread);
+
+		for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+			if ((data->cpu_mask & BIT(i)) == 0) {
+				k_thread_cpu_mask_disable(data->thread, i);
+			} else {
+				k_thread_cpu_mask_enable(data->thread, i);
+			}
+		}
```
Comment (evgeniy-paltsev) [inline tests/arch/arc/arc_vpx_lock/src/main.c:79]: Hmm, it looks like over-complicated way to do `k_thread_cpu_pin`.

---

Moreover the `test_helper_entry` itself looks over-complicated (or I'm missing some test logic here).
If we just need to pin a test thread to one CPU I'd just do it this way (and drop `test_helper_entry` entirely):

```
ZTEST(vpx_lock, test_case)
{
        k_thread_create(&test_payload, ...);
        k_thread_cpu_pin(&test_payload, cpu_id);
        k_thread_start(&test_payload);
}
```

**Comment by peter-mitsis** [comment]
@ruuddw - Why couple it to the ARC_DSP feature? Judging by the question, I think I misunderstood something / made a bad mental connection in thinking that the VPX registers were tied to the DSP registers. I can correct that.

Why not a more symmetrical interface? I've been waffling on that for a bit. One of my concerns was that if the thread that has the VPX lock is aborted before it releases that semaphore, then no other thread would be able to use the VPX registers as the semaphore would never get unlocked. To work around that, I wanted a mechanism that could in theory allow some sort of monitoring thread/ISR to force their availability. If such a feature is wanted, perhaps it would be better to have a separate function--something like arc_vpx_unlock_force(uint32_t cpu_id) and then we could otherwise keep the symmetry.

Are these routines allowed from an ISR? They are supposed to be. The intent is to only check the cpu_mask if it is called by a thread (_not_ in an ISR). If it is in an ISR, then it will always only execute on the current CPU; otherwise it is a thread, and we want to verify that it is not going to get scheduled to execute on another CPU. There is a potential hole in that the application developer change the set of CPUs that the thread could execute on after this check, but there is a documentation note saying not to do that.

File: tests/arch/arc/arc_vpx_lock/src/main.c (hunk 1 above)
Comment (peter-mitsis) [reply tests/arch/arc/arc_vpx_lock/src/main.c:79]: This is a leftover artefact of previous implementations and test versions when the cpu_mask check in arc_vpx_lock() was an if-statement and not an assert. I will simplify it.

**Comment by peter-mitsis** [comment]
I decoupled the VPX from DSP and did some additional cleanup based on the feedback thus far.

**Comment by ruuddw** [comment]
Hi Peter, thanks for the explanations and rationale. I now understand the recovery unlocking idea, makes sense to have it as a separate function. The alternative would have been to add an id input param to both lock and unlock, instead of relying on _current_cpu->id. That leaves more to the application, but as the solution already has assumptions on well-behaving applications I think that should be fine.
Note that the ARC Connect IDU (multicore interrupt distribution unit) can route interrupts to different CPUs -> just like for tasks, some pinning for ISRs might be needed when relying on using vector instructions in interrupt context. And when unlocking from ISR context.

**Comment by peter-mitsis** [comment]
@ruddw - Regarding the interrupt routing, I think you are saying that ...

1. The ARC Connect IDU can pick which CPU will service the interrupt. One implication of this is that it may lead to unpredictable failures (to lock) in the ISR if the servicing CPU executes threads that lock the VPX registers. This may be further exacerbated if nested interrupts come into play.
2. The application developer can ameliorate this somewhat by pinning interrupts to specific CPUs to help create a system whose behavior can be more easily predicted and modeled.

**Comment by peter-mitsis** [comment]
@ruuddw - I updated the PR so that neither arc_vpx_lock() nor arc_vpx_unlock() can be called from an ISR. However that restriction does not exist with arc_vpx_unlock_force().

**Review by ruuddw — APPROVED** [review]

**Comment by ruuddw** [comment]
> I updated the PR so that neither arc_vpx_lock() nor arc_vpx_unlock() can be called from an ISR. However that restriction does not exist with arc_vpx_unlock_force().

Thanks! This makes all sense to me now, just approved the PR.

**Review by nashif — APPROVED** [review]

**Pull request merged**